File: tourism/src/main/java/com/tashin/tourism/destination/entity/Destination.java
package com.tashin.tourism.destination.entity;

import java.time.Instant;
import java.util.UUID;

import org.locationtech.jts.geom.Point;

import com.tashin.tourism.region.entity.Region;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Entity
@Table(name = "destinations")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Destination {

    @Id
    @GeneratedValue
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "region_id")
    private Region region;

    @Column(nullable = false, length = 200)
    private String name;

    @Column(nullable = false, unique = true, length = 220)
    private String slug;

    @Column(name = "short_description", columnDefinition = "TEXT")
    private String shortDescription;

    @Column(columnDefinition = "TEXT")
    private String description;

    /**
     * PostGIS geography(Point,4326) column.
     * JTS Point: x = longitude, y = latitude.
     */
    @Column(columnDefinition = "geography(Point,4326)", nullable = false)
    private Point location;

    @Column(nullable = false)
    private boolean featured;

    @Column(nullable = false, length = 30)
    private String status;

    @Column(name = "seo_title", length = 255)
    private String seoTitle;

    @Column(name = "seo_description", columnDefinition = "TEXT")
    private String seoDescription;

    @Column(name = "created_at")
    private Instant createdAt;

    @Column(name = "updated_at")
    private Instant updatedAt;

    @Column(name = "deleted_at")
    private Instant deletedAt;

    @Column(name = "published_at")
    private Instant publishedAt;

    @Column(name = "published_by")
    private UUID publishedBy;

    @Column(name = "rejection_reason", columnDefinition = "TEXT")
    private String rejectionReason;

    @PrePersist
    void onCreate() {
        Instant now = Instant.now();
        createdAt = now;
        updatedAt = now;
    }

    @PreUpdate
    void onUpdate() {
        updatedAt = Instant.now();
    }
}

File: sri-lanka-tourism/src/types/destination.ts
export interface LocationResponse {
  latitude: number;
  longitude: number;
}

export interface RegionSummary {
  id: string;
  name: string;
  slug: string;
}

export interface Destination {
  id: string;
  name: string;
  slug: string;
  shortDescription: string;
  description: string;
  /** Nested location object from backend */
  location?: LocationResponse;
  /** Flat fields (used in some DTOs) */
  latitude?: number;
  longitude?: number;
  region: RegionSummary | null;
  featured: boolean;
  status: string;
  seoTitle?: string;
  seoDescription?: string;
}

/** Helper — resolves latitude regardless of response shape */
export function getLatitude(d: Destination): number {
  return d.location?.latitude ?? d.latitude ?? 0;
}

export function getLongitude(d: Destination): number {
  return d.location?.longitude ?? d.longitude ?? 0;
}
